### DISCUSSION TITLE
How to use WiFi on particle argon using esp32

### DISCUSSION THREAD

**Discussion by vinilima07** [zephyrproject-rtos/zephyr#29845] [Q&A]
Hi, i am new on Zephyr. 
I have an Argon Nordic Semiconductor nRF52840 SoC with  ESP32 Wifi Coprocessor and i am trying to create communication beween then.  

Argon Datasheet: https://docs.particle.io/datasheets/wi-fi/argon-datasheet/

I have already established the following network configuration for prj.conf
```
CONFIG_NETWORKING=y
CONFIG_WIFI=y
CONFIG_NET_IPV4=y
CONFIG_WIFI_ESP_AT_VERSION_2_0=y
CONFIG_WIFI_ESP=y
CONFIG_NET_L2_WIFI_MGMT=y
```
However, I don't know where to start so that I can issue AT commands to esp32. Sorry for anything but I am very confused with the documentation, so I had to resort here.

**Comment by whati001** [comment]
Hi @vinilima07 

If the esp32 is connected via UART, you can use the Zephyr [NetworkManager](https://docs.zephyrproject.org/latest/connectivity/networking/api/net_mgmt.html) in combination with the "[espressif,esp-at](https://docs.zephyrproject.org/latest/build/dts/api/bindings/wifi/espressif%2Cesp-at.html)" driver.

The driver define a network device which allows you establish a L2 wifi connection with your AP. After that you can use the normal [socket](https://docs.zephyrproject.org/latest/connectivity/networking/api/sockets.html) stack to establish a TCP/UDP connection and further a HTTP request or something else.

Please have a look at the [WiFi-Shell](https://docs.zephyrproject.org/latest/samples/net/wifi/README.html) example. You have to connect the esp32 properly to the nrf uart port and it as subnode. Please find below an example of the devicetree configuration with hardware control flow enabled:
```dts
 &uart1_default {
	cts-rts {
		psels = <NRF_PSEL(UART_CTS, 1, 3)>,
			<NRF_PSEL(UART_RTS, 1, 4)>;
	};
};
&uart1_sleep {
	cts-rts {
		psels = <NRF_PSEL(UART_CTS, 1, 3)>,
			<NRF_PSEL(UART_RTS, 1, 4)>;
		low-power-enable;
	};
};
&uart1 {
	status = "okay";
	hw-flow-control;

	esp_wifi: esp-wifi {
		compatible = "espressif,esp-at";
		// power-gpios = <&gpio1 5 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
		status = "okay";
	};
};
```
The final connection should be as follows:
| Pin | Nordic | ESP32 |
| --- | ------ | ----- |
| RX  | P1.01  | 17    |
| TX  | P1.02  | 16    |
| CTS | P1.03  | 14    |
| RTS | P1.04  | 15    |

After that, you can use the network manager to perform a scan (`NET_REQUEST_WIFI_SCAN`), connect to an AP (`NET_REQUEST_WIFI_CONNECT`) and the socket api (`socket` and `connect`) to establish a connection.

You can also have a look at this (private) example I have written some time ago to mess around with the esp32 at stack: https://github.com/whati001/zephyr_playground/tree/main/esp32_wifi_http
